include/settings/settings.h

```diff
@@ -120,6 +131,15 @@ int settings_register(struct settings_handler *cf);
  */
 int settings_load(void);
 
+/**
+ * Load limited set of serialized items from registered persistence sources.
+ * Handlers for serialized item subtrees registered earlier will be called for
+ * encountered values that belong to the subtree
+ *
+ * @return 0 on success, non-zero on failure.
+ */
+int settings_load_subtree(const char *subtree);
```

Comment:
Is this something I could call on bt_init? What if the application is also calling settings_load by itself then? Would that cause the handler to be called once again? I guess we should clearly state on the documentation this sort of behavior.

Reply:
Yes you could, the mutex would not stop the handler to be called again. For the documentation it should be clear:
settings_load(): loads all entries,
settings_load_subtree(): loads limited set of entries (subtree is not limited to handler names, it can be a subset of a handler name).

But maybe your are asking something else: I have a handler that I want to call myself, and nobody else, how do I do this ?
Option 1: Add a flag to the set routine and only allow processing of the set routine if this flag is set. Do the loading by: enabling the flag, call settings_load_subtree, disable the flag. This is not thread-safe but might be sufficient.
Option 2: Don't register the settings_handler yet, do the loading by register the handler, call settings_load_subtree, deregister the handler. At the moment this is not yet thread-safe, but this could be converted into one routine and made thread-safe.

Reply:
My question was that if this would cause the handler to be called multiple times, one by load and another by load_subtree that may cause extra loading from no reason, I guess the idea is that each subsystem calls load_substree and load would take care of the remaining settings which most likely are just app specific, this way the more natural sequence would be subsystem blah settings_load_subtree("blah") (marked it as loaded)... -> app settings_load()

Reply:
oops, sorry forgot to answer the question: Yes, it would be called several times.

There is no requirement on how to use it, the idea is that it would be needed to call a settings_load_subtree() as a result of a connection request so that only limited RAM would be occupied for storing this information.

Your natural sequence could also be done: using option 2 from above settings_load_handler(&blah_handler): registers blah, loads blah elements, unregister blah (this is the marked as loaded). Do this for each subsystem and finally call settings_load for all app related stuff.

Should I add a settings_load_handler() like defined above ?

Reply:
It depends on how we wish the system to be designed. I believe that the full settings load should be called only once - on system start. And it should be left to the module implementation to decide how we wish to tread incoming data (for example, any CCCD data may be ignored if we do not have an active connection already).

Then when the system works, we can call settings_load_subtree (or load_selected from #16609) whenever needed.

Reply:
@Laczen 
> Should I add a settings_load_handler() like defined above ?

I would say rather not. If it would really be required - leave it for another PR.

Reply:
^^ for me adding a flag for disabling a handler sounds well (it might be keep in separate structure as we still can put handlers declaration to NVM). I prefer to have this in another PR - scope of this one is enough and I'm not sure what will be the best solution (maybe list of sub-trees to be ignored?). No doubts that "per implementation enabling" is always possible.

Reply:
> It depends on how we wish the system to be designed. I believe that the full settings load should be called only once - on system start. And it should be left to the module implementation to decide how we wish to tread incoming data (for example, any CCCD data may be ignored if we do not have an active connection already).

That makes the assumption that setting_load is already the very first thing the application do but we end up having bugs with that as you may find warning such as this:

https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/hci_core.c#L4706

So far we don't have a proper dependecy handling of the subsystem that do require setting_load to be called before or after it, most should be after I suppose but we need a way to define that.

Reply:
@rakons, @nvlsianpu I agree with @Vudentz, settings cannot dictate how or when `settings_load()` or `settings_load_subtree()/settings_load_handler()` should be called. Settings cannot make any assumption on when it will be called.

@Vudentz, I don't think settings should provide any solution for module dependencies, the modules need to handle this themselves, but for this to work you need to be able to determine that something has been done or have the possibility to do something. This should be possible with a `settings_load_handler()` routine.

@rakons, @nvlsianpu, regarding the step to flash, the routine `settings_register()/settings_deregister()` would disappear and routines dependent on these would need to be reworked. The reason to do this might have disappeared completely when using the `settings_deregister()` method, the RAM usage of settings can be kept very low by deregistering handlers.

Reply:
@Laczen - I believe you were saying something about simplifying the whole thing. Now you are providing another way of complication and another step when moving aside from RAM implementation.

I will repeat it - Do not think about the settings as only the way to store and restore data. If we wish to do it - we would go to pull architecture. The subtree or pattern loading is provided here to allow some behaviour similar to what we can get from pull architecture but sticking to the way it is handled in modules.

Dynamic registering and deregistering handlers add another level of complexity where modules decide whatever it wishes to be handled now or not - now imagine how to pair it with settings injection from CLI or from the Ethernet.

Reply:
Why do we even need a settings_load()? Couldn't each module just call settings_load_subtree() when appropriate on init (or on demand)? The app can always have its own subtree (fx. "app/") and use settings_load_subtree().

Reply:
@fnde-ot, completely agree, we could remove the settings_load() completely. Combining this with the remark from @Vudentz in #16676 would open a lot of opportunities. There of course would be
a disadvantage that all entries in the backend are read multiple times, but maybe even this can be avoided (e.g. for the bluetooth system): let us suppose there are multiple subsystems defined (bt/btmesh, bt/ccc, bt/gatt ... these are just examples I'm not a bt expert) each of these would define their own set routines. To load all bt related settings we would define bt_set routines as:
```
int bt_set(const char *name, ...) {
  char *next;
#ifdef(BTMESH)
  if (settings_name_cmp(name,"btmesh", &next)) {
    return btmesh_set(next, ...);
  }
#endif
#ifdef(BTCCC)
  if (settings_name_cmp(name, "ccc", &next)) {
    return ccc_set(next, ....);
  }
#endif
}

int bt_load() {
  settings_load_subtree("bt", bt_set);
}
```

Reply:
And you cannot agree with my proposal as it requires too much of rework inside modules. I ask you not to go that direction here - separate this idea and offer it in another PR.

Reply:
@Laczen @rakons, Sounds to me more like a backend problem. On-demand or subtree loading should not require significant changes to existing modules, appart from adding calls to settings_load_subtree(...) in the appropriate init functions.

The multiple-read would not be an issue if the backend properly separated settings from different modules (not a flash expert, but maybe using multiple files or grouping settings from a single module together and reading from a specific offset). But then again, this could require to rethink the way existing backends work, which would be out of scope for this PR.

We can still remove settings_load from this PR and rework the backends in another one.

Reply:
@rakons, I cannot agree with your proposal because:
a. The pattern search is to much code and not needed
b. The split up to get a dynamic generated name is not done and is relying on string conversion routines to do the split up. This only works if there is a conversion routine available.

I have accepted doing the rework inside the modules by not introducing the setiings_name_separate(), but there is a very specific reason for this: reducing RAM usage in the case where a dynamic name would be used at the top of some tree where most entries under this would have a fixed part (e.g. `bt/ccc/0xfffe/info/data/val1, bt/ccc/0xffe/info/data/val2, bt/ccc/0xffe/id`, and the same entries for `bt/ccc/0xfffd...`).

This however has nothing to do with the discussion above: the proposal has triggered other (new) ideas to make the proposal even simpler (remove registering of handlers and looking up handlers) and it would be a pity not to take these ideas into account.